### PR TITLE
Add masks to wandblogger.log_image and cometmllogger.log_image and refactor ImageVisualizer to use log_image [WIP]

### PR DESCRIPTION
**Pull request by eracah**
# What does this PR do?

Adds in mask support for `WandBLogger.log_image` and refactors ImageVisualizer.

Manual tests

### Inputs
[wandb](https://wandb.ai/mosaic-ml/evan-tb-test/runs/16pxb4vr?workspace=user-evan-mosaicml)
[comet_ml](https://www.comet.com/mosaic-ml/general/55be87c9363b43279f9ddd4072424e9d?experiment-tab=images)

### Segmentation
[comet](https://www.comet.com/mosaic-ml/daniel-test/9093e475c5d547d2a246f65cff75a2ab?experiment-tab=images)
[wandb](https://wandb.ai/mosaic-ml/daniel-debug/runs/1mz083fu?workspace=user-danielking)

# What issue(s) does this change relate to?

solves CO-682 and CO-1361

# Before submitting
- [x] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [x] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [ ] Did you update any related docs and document your change?
- [x] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [x] Did you run the tests locally to make sure they pass?
- [x] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

File: tests/loggers/test_cometml_logger.py [hunk 1]
```diff
@@ -76,16 +77,42 @@ def test_comet_ml_log_image_saves_images(images: torch.Tensor, channels_last: bo
...
+    'pred': torch.randint(0, 10, (4, 32, 32)),
+    'pred2': torch.randint(0, 10, (4, 32, 32))
+})])
+def test_comet_ml_log_image_saves_images_with_masks(images: torch.Tensor, masks: Dict[str, torch.Tensor],
+                                                    comet_logger: CometMLLogger, comet_offline_directory: str):
+    assert isinstance(comet_offline_directory, str)
+    # Count expected images and generate numpy arrays from torch tensors.
+    num_masks = len(masks.keys())
+    num_images = images.shape[0] if images.ndim == 4 else 1
+    expected_num_images = num_images * (2 * num_masks + 1)
```
Comment (dakinggg) [inline tests/loggers/test_cometml_logger.py:96]: Why is this the expected number of images?

Comment (eracah) [reply tests/loggers/test_cometml_logger.py:96]: For each image, we plot the image (num_images * 1) + the image with the mask overlaid and the mask by itself for each mask: (num_images * 2 * num_masks).

File: tests/loggers/test_cometml_logger.py [hunk 2]
```diff
@@ -76,16 +77,42 @@ def test_comet_ml_log_image_saves_images(images: torch.Tensor, channels_last: bo
...
-    'images,channels_last',
-    [
-        (torch.rand(32, 0), False),  # Has zero in dimension.
-        (torch.rand(4, 4, 8, 32, 32), False),  # > 4 dim.
-        ([torch.rand(4, 32, 32, 3)], True),
-    ])  # sequence > 3 dim.
-def test_comet_ml_log_image_errors_out(comet_logger, images, channels_last):
-    with pytest.raises(ValueError):
-        comet_logger.log_images(images, channels_last=channels_last)
+@pytest.mark.parametrize('images,masks', [(torch.randint(0, 256, (32, 32, 3)), {
```
Comment (dakinggg) [inline tests/loggers/test_cometml_logger.py:80]: Can you add a test with not channels last as well?

File: tests/loggers/test_wandb_logger.py [hunk 3]
```diff
@@ -76,6 +76,37 @@ def test_wandb_ml_log_image_errors_out(test_wandb_logger, images, channels_last)
         test_wandb_logger.log_images(images, channels_last=channels_last)
 
 
+@pytest.mark.parametrize('images,masks', [(torch.randint(0, 256, (32, 32, 3)), {
```
Comment (dakinggg) [inline tests/loggers/test_wandb_logger.py:79]: Same, add a test without channels last

File: composer/loggers/wandb_logger.py [hunk 4]
```diff
@@ -335,4 +354,36 @@ def _convert_to_wandb_image(image: Union[np.ndarray, torch.Tensor], channels_las
     if not channels_last:
         assert isinstance(image, np.ndarray)
         image = image.transpose(1, 2, 0)
-    return image
+    return np.asarray(image)
```
Comment (dakinggg) [inline composer/loggers/wandb_logger.py:357]: I think this conversion should not be necessary because the top of this function converts it to numpy?

File: composer/loggers/wandb_logger.py [hunk 5]
```diff
@@ -335,4 +354,36 @@ def _convert_to_wandb_image(image: Union[np.ndarray, torch.Tensor], channels_las
     if not channels_last:
```
Comment (dakinggg) [inline composer/loggers/wandb_logger.py:354]: This section is causing a test to fail (and causes the image visualizer to fail when i run it locally).

**Review by dakinggg — COMMENTED** [review]
Will rereview once the tests are passing. I think the tests are failing are due to 1) a bad reshape in wandb (mentioned in an inline comment), and needing to implement the log_images function for the InMemoryLogger. Also, could you add an ImageVisualizer+Comet test, similar to the one for WandB?

File: composer/callbacks/image_visualizer.py [hunk 6]
```diff
@@ -171,26 +149,24 @@ def eval_after_forward(self, state: State, logger: Logger):
...
 def _make_segmentation_images(inputs: torch.Tensor,
                               targets: torch.Tensor,
-                              outputs: torch.Tensor,
+                              outputs: Union[torch.Tensor, Sequence[torch.Tensor]],
                               num_images: int,
                               channels_last: bool = False):
-    import wandb
+    if isinstance(outputs, Sequence):
+        outputs = torch.stack(list(outputs))
+    # import wandb
```
Comment (dakinggg) [inline composer/callbacks/image_visualizer.py:169]: remove commented out code

File: tests/loggers/test_cometml_logger.py (hunk 1 above)
Comment (eracah) [reply tests/loggers/test_cometml_logger.py]: done

File: composer/loggers/wandb_logger.py [hunk 7]
```diff
@@ -335,4 +354,36 @@ def _convert_to_wandb_image(image: Union[np.ndarray, torch.Tensor], channels_las
...
+    for mask_name, mask_data in masks.items():
+        if not isinstance(mask_data, Sequence):
+            mask_data = mask_data.squeeze()
+            if mask_data.ndim == 2:
+                mask_data = [mask_data]
+        preprocesssed_masks[mask_name] = np.stack([_convert_to_wandb_mask(mask, channels_last) for mask in mask_data])
+    return preprocesssed_masks
+
+
+def _create_wandb_masks_generator(masks: Dict[str, Union[np.ndarray, torch.Tensor, Sequence[Union[np.ndarray,
```
[the remark this answers is not in the document]
Comment (eracah) [reply composer/loggers/wandb_logger.py:380]: It should be good because it converts Tensors to numpy arrays in `_convert_to_wandb_image` and everything has to go through that function.

File: tests/loggers/test_cometml_logger.py (hunk 2 above)
Comment (eracah) [reply tests/loggers/test_cometml_logger.py]: done

File: composer/callbacks/image_visualizer.py (hunk 6 above)
Comment (eracah) [reply composer/callbacks/image_visualizer.py:166]: done

File: tests/loggers/test_wandb_logger.py (hunk 3 above)
Comment (eracah) [reply tests/loggers/test_wandb_logger.py]: done

File: composer/loggers/wandb_logger.py (hunk 5 above)
Comment (eracah) [reply composer/loggers/wandb_logger.py:354]: fixed?

File: composer/loggers/wandb_logger.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (dakinggg) [reply composer/loggers/wandb_logger.py:379]: The masks dict does not go through `_convert_to_wandb_image` I think? So I think right now if you actually passed a torch tensor as the masks to `WandBLogger.log_images` I think it would crash. It is mostly a non issue because your code is the only caller of this function, but it would be good to get the types right here.

File: composer/callbacks/image_visualizer.py [hunk 8]
```diff
@@ -112,40 +110,20 @@ def __init__(self,
...
-        if not isinstance(targets, torch.Tensor):
-            raise NotImplementedError('Multiple target tensors not supported yet')
-        if not isinstance(outputs, torch.Tensor):
-            raise NotImplementedError('Multiple output tensors not supported yet')
-
-        table = _make_segmentation_images(inputs, targets, outputs, self.num_images, self.channels_last)
+        assert isinstance(outputs, torch.Tensor)
+
+        images, masks = _make_segmentation_images(inputs, targets, outputs, self.num_images, self.channels_last)
         # Only log to the wandb logger if it is available
```
Comment (dakinggg) [inline composer/callbacks/image_visualizer.py:125]: Delete out of date comment

File: composer/callbacks/image_visualizer.py [hunk 9]
```diff
@@ -171,26 +149,24 @@ def eval_after_forward(self, state: State, logger: Logger):
 
 
 def _make_input_images(inputs: torch.Tensor, num_images: int, channels_last: bool = False):
-    import wandb
+    #import wandb
```
[the remark this answers is not in the document]
Comment (dakinggg) [reply composer/callbacks/image_visualizer.py:152]: Seems like this is still here?

File: composer/callbacks/image_visualizer.py (hunk 6 above)
Comment (dakinggg) [reply composer/callbacks/image_visualizer.py:166]: Seems like this is still here?

File: composer/callbacks/image_visualizer.py [hunk 10]
```diff
@@ -171,32 +149,26 @@ def eval_after_forward(self, state: State, logger: Logger):
 
 
 def _make_input_images(inputs: torch.Tensor, num_images: int, channels_last: bool = False):
```
Comment (dakinggg) [inline composer/callbacks/image_visualizer.py:151]: seems like the `channels_last` arg is no longer needed here?

File: examples/segmentation/train_deeplabv3_ade20k.py [hunk 11]
```diff
@@ -328,6 +330,9 @@ def combo_loss(output, target):
         logger = WandBLogger(entity=args.wandb_entity, project=args.wandb_project)
         logging.info('Built Weights and Biases logger')
 
+    callbacks = [speed_monitor, lr_monitor, checkpoint_saver]
+    if args.image_viz:
+        callbacks.append(ImageVisualizer())
```
Comment (dakinggg) [inline examples/segmentation/train_deeplabv3_ade20k.py:335]: I guess its ok to just use default args here, but should probably set mode to segmentation?

File: tests/callbacks/test_image_visualizer.py [hunk 12]
```diff
@@ -1,112 +1,235 @@
...
-try:
-    import wandb
-    _WANDB_INSTALLED = True
-    del wandb  # unused
-except ImportError:
-    _WANDB_INSTALLED = False
-
-
-@pytest.mark.skipif(not _WANDB_INSTALLED, reason='Wandb is optional')
-@pytest.mark.parametrize('interval', ['9ba', '90ba', '2ep', '3ep', '7ep'])
```
Comment (dakinggg) [inline tests/callbacks/test_image_visualizer.py:26]: Can you add the interval parametrization back to at least one of the tests? There was a bug related to this, so I don't want to regress coverage for that.

File: tests/callbacks/test_image_visualizer.py [hunk 13]
```diff
@@ -1,112 +1,235 @@
...
-        max_duration=f'{num_train_epochs}ep',
-    )
+from tests.common.datasets import RandomImageDataset, RandomSegmentationDataset
+from tests.common.models import SimpleConvModel, SimpleSegmentationModel
+
+
+@pytest.fixture
+def test_wandb_logger(tmp_path, dummy_state):
+    pytest.importorskip('wandb', reason='wandb is optional')
+    os.environ['WANDB_DIR'] = str(tmp_path)
```
Comment (dakinggg) [inline tests/callbacks/test_image_visualizer.py:24]: I'm not sure about env var setting in a specific test fixture, because this sets them for all tests, which may cause unexpected results across tests. I'd rather either add whatever WandB env vars you want to the autoused fixture, or unset these env vars at the end of the test. Perhaps a minor thing, not sure what you think.

File: tests/callbacks/test_image_visualizer.py [hunk 14]
```diff
@@ -1,112 +1,235 @@
...
+    # delete trainer to force WandBLogger to clean up in post_close
+    del trainer
+
+    expected_number_train_images = (batch_size * max_duration) / image_interval
+    expected_number_eval_images = (max_duration / eval_interval) * batch_size
 
-    assert isinstance(image_visualizer.interval, Time)
-    assert num_train_tables == expected_train_tables
-    assert num_eval_tables == num_train_epochs
+    # WandB Images are stored in cache when a WandB table is used.
```
Comment (dakinggg) [inline tests/callbacks/test_image_visualizer.py:80]: It looks like the segmentation tests passes without going to the cache? It would be nice if this test and the segmentation test didn't have to look in different places for the images. Is that possible, or did you determine that looking in the cache here is required?

File: tests/loggers/test_wandb_logger.py [hunk 15]
```diff
@@ -76,6 +85,113 @@ def test_wandb_ml_log_image_errors_out(test_wandb_logger, images, channels_last)
...
+    image_count = 0
+    for image_table_file in image_table_files:
+        table_columns = json.load(open(image_table_file.absolute()))['data']
+        num_images = sum([1 for column in table_columns if column[0] == 'Image'])
+        image_count += num_images
+
+    assert image_count == expected_num_images
+
+
+def test_wandb_and_image_visualizer(tmp_path, test_wandb_logger):
```
Comment (dakinggg) [inline tests/loggers/test_wandb_logger.py:155]: Are the image visualizer tests in the test logger files or in the test image visualizer file? It seems like its in both places right now?

File: composer/loggers/wandb_logger.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (dakinggg) [reply composer/loggers/wandb_logger.py:379]: ah nevermind, i see it now

**Review by dakinggg — COMMENTED** [review]
Basically looks good to me, would still like to see a manual test since these are images and just want to double check that they end up properly formed in the UI before approving :)

File: tests/loggers/test_wandb_logger.py [hunk 16]
```diff
@@ -39,11 +43,16 @@ def test_wandb_logger(tmp_path, dummy_state):
     return wandb_logger
 
 
+def teardown_function(function):
```
Comment (dakinggg) [inline tests/loggers/test_wandb_logger.py:46]: Should there be a teardown in test comet and test image visualizer also?

File: tests/callbacks/test_image_visualizer.py (hunk 13 above)
Comment (dakinggg) [reply tests/callbacks/test_image_visualizer.py:24]: I'm ok with this if you think its ok

File: composer/callbacks/image_visualizer.py (hunk 6 above)
Comment (eracah) [reply composer/callbacks/image_visualizer.py]: done?

File: composer/callbacks/image_visualizer.py (hunk 10 above)
Comment (eracah) [reply composer/callbacks/image_visualizer.py]: done

File: examples/segmentation/train_deeplabv3_ade20k.py (hunk 11 above)
Comment (eracah) [reply examples/segmentation/train_deeplabv3_ade20k.py]: fixed

File: tests/loggers/test_wandb_logger.py (hunk 15 above)
Comment (eracah) [reply tests/loggers/test_wandb_logger.py]: just misnamed

File: tests/callbacks/test_image_visualizer.py [hunk 17]
```diff
@@ -1,112 +1,283 @@
...
+    from composer.loggers import CometMLLogger
+
+    # Set offline directory.
+    os.environ['COMET_OFFLINE_DIRECTORY'] = comet_offline_directory
+
+    comet_logger = CometMLLogger()
+    return comet_logger
+
+
+# @pytest.mark.parametrize('interval', ['9ba', '90ba', '2ep', '3ep', '7ep'])
```
Comment (dakinggg) [inline tests/callbacks/test_image_visualizer.py:54]: remove commented test

**Review by dakinggg — APPROVED** [review]
